eclipse-che/che-operator
# feat: Support node selector and pod tolerations in devworkspaces

### What does this PR do?

This adds support for a couple of things:

* adds support for workspace pod tolerations and node selector as properties in
  the `CheCluster` CRD v1 under `spec.server.workspacePodTolerations` and `spec.server.workspacePodNodeSelector`
  in addition to the `CHE_WORKSPACE_POD_NODE__SELECTOR` and `CHE_WORKSPACE_POD_TOLERATIONS__JSON` custom
  properties,
* adds support for the same as properties in the CRD v2alpha1 under `spec.workspaces.podTolerations` and 
  `spec.workspaces.podNodeSelector`,
* implements conversion between the two (when converting to v2alpha1, the values are read out of either
  the properties, or if not found there, from the custom che properties)
* the usernamespace reconciler now annotates the user namespace/project with annotations understood by
  the devworkspace operator to apply the node selector and pod toleration configuration to the actual 
  devworkspace pods.

### Screenshot/screencast of this PR
none

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/20884

### How to test this PR?
1. Install the che-operator from this PR (I've pre-built an image `quay.io/lkrejci/che-operator:issue-20884`)
1. Configure node selector and/or pod tolerations in the CheCluster CR
1. Create a devworkspace
1. The configuration should be applied to the workspace pod

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.
